pkg/cluster/config/config_types_test.go

```diff
@@ -0,0 +1,189 @@
+			ExpectBootStrapControlPlane:  utilpointer.StringPtr("controlplane"),
+			ExpectSecondaryControlPlanes: nil,
+			ExpectWorkers:                []string{"worker"},
+			ExpectError:                  false,
+		},
+		{
+			TestName: "Full HA cluster", // NB. This test case test that provisioning order is applied to all the node lists as well
+			Nodes: []*Node{
+				{Role: WorkerRole},
+				{Role: ControlPlaneRole},
+				{Role: ExternalEtcdRole},
+				{Role: ControlPlaneRole},
+				{Role: WorkerRole},
+				{Role: ControlPlaneRole},
+				{Role: ExternalLoadBalancerRole},
+			},
+			ExpectNodes:                  []string{"etcd", "lb", "controlplane1", "controlplane2", "controlplane3", "worker1", "worker2"},
+			ExpectControlPlanes:          []string{"controlplane1", "controlplane2", "controlplane3"},
+			ExpectBootStrapControlPlane:  utilpointer.StringPtr("controlplane1"),
+			ExpectSecondaryControlPlanes: []string{"controlplane2", "controlplane3"},
+			ExpectWorkers:                []string{"worker1", "worker2"},
+			ExpectEtcd:                   utilpointer.StringPtr("etcd"),
+			ExpectLoadBalancer:           utilpointer.StringPtr("lb"),
+			ExpectError:                  false,
+		},
+		{
+			TestName: "Fails because two etcds Nodes are added",
+			Nodes: []*Node{
+				{Role: ExternalEtcdRole},
+				{Role: ExternalEtcdRole},
+			},
+			ExpectError: true,
+		},
+		{
+			TestName: "Fails because two load balancer Nodes are added",
+			Nodes: []*Node{
+				{Role: ExternalLoadBalancerRole},
+				{Role: ExternalLoadBalancerRole},
+			},
+			ExpectError: true,
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.TestName, func(t2 *testing.T) {
+			// Adding Nodes to the config until first error or completing all Nodes
+			var cfg = Config{}
+			var err error
+			for _, n := range c.Nodes {
+				if e := cfg.Add(n); e != nil {
+					err = e
+					break
+				}
+			}
+			// the error can be:
+			// - nil, in which case we should expect no errors or fail
+			if err != nil {
+				if !c.ExpectError {
+					t2.Fatalf("unexpected error while adding Nodes: %v", err)
+				}
+				return
+			}
+			// - not nil, in which case we should expect errors or fail
+			if err == nil {
+				if c.ExpectError {
+					t2.Fatalf("unexpected lack or error while adding Nodes")
+				}
+			}
+
+			// Fail if Nodes does not match
+			checkNodeList(t2, cfg.Nodes(), c.ExpectNodes)
+
+			// Fail if fields derived from Nodes does not match
+			checkNodeList(t2, cfg.ControlPlanes(), c.ExpectControlPlanes)
+			checkNode(t2, cfg.BootStrapControlPlane(), c.ExpectBootStrapControlPlane)
+			checkNodeList(t2, cfg.SecondaryControlPlanes(), c.ExpectSecondaryControlPlanes)
+			checkNodeList(t2, cfg.Workers(), c.ExpectWorkers)
+			checkNode(t2, cfg.ExternalEtcd(), c.ExpectEtcd)
+			checkNode(t2, cfg.ExternalLoadBalancer(), c.ExpectLoadBalancer)
+		})
+	}
+}
+
+func checkNode(t *testing.T, n *Node, name *string) {
+	if (n == nil) != (name == nil) {
+		t.Errorf("expected %v node, saw %v", name, n)
```

Comment:
isn't it the same as doing `n != name` ?

also not sure how to formulate this message.
ideally we should be comparing objects of the same type e.g. bool vs bool.

Reply:
n != name can't be used because n and name are different types

the test is validate node ordering, and checking only the node name is a way to do this and at the same time keep test case definition simple and readable